### PR TITLE
Fixes 14722: add network policy generator/tester

### LINKED ISSUE #14722: Feature Request: use mattfenwick/cyclonus to test network policies
As [discussed on Slack](https://cilium.slack.com/archives/C2B917YHE/p1611353849014400) with @mattfenwick and @aanm .

## Proposal / RFE

**Is your feature request related to a problem?**

https://github.com/mattfenwick/cyclonus is a sophisticated tool for testing Kubernetes `NetworkPolicy`s, and has already found a bug in Cilium's implementation (#14678). It would be fantastic to integrate cyclonus into Cilium's CI to catch problems as early as possible, and to benefit from the rich range of tests in cyclonus.

**Describe the solution you'd like**

As a first step, integrate cyclonus into Cilium's CI. It is possible that some tests will fail (e.g. as identified by #14678) so a mechanism will be needed to mark such failures as expected until Cilium is fixed. As cyclonus expands its range of tests, it might hit expected failures due to, for example [missing functionality in Cilium's implementation of `NetworkPolicy`s](https://docs.cilium.io/en/v1.9/concepts/kubernetes/policy/#k8s-policy).

### PR DESCRIPTION
**Pull request by mattfenwick**
TODOs before merging:

 - [x] read the output from the job to determine success/failure
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Handle TODOs in the code

Fixes: #14722 

```release-note
Add cyclonus network policy tester.
```

## Future plans

This is a PoC -- the tests run cover some network policy features but not all.  The goal of this PR is just to get the ball rolling with Cyclonus: make sure the tests are stable and work as expected.

When future Cyclonus releases include additional tests, it'll be an easy update -- we can bump the version in the Cyclonus job yaml to pull in those new tests.

### CODE REVIEW

**Comment by mattfenwick** [comment]
Putting this WIP up to start figuring out what the job needs to do!  Probably doesn't work yet 😄

**Comment by mattfenwick** [comment]
Removed the 'check prometheus metrics' step because it caused the cyclonus job to fail, which doesn't seem to be related to anything cyclonus is doing:

```
package golang.org/x/oauth2: unrecognized import path "golang.org/x/oauth2": reading https://golang.org/x/oauth2?go-get=1: 502 Bad Gateway
package golang.org/x/oauth2/google: unrecognized import path "golang.org/x/oauth2/google": reading https://golang.org/x/oauth2/google?go-get=1: 502 Bad Gateway
Error: Process completed with exit code 1.
```

**Comment by mattfenwick** [comment]
Successful test run to make sure this is working: https://github.com/cilium/cilium/pull/14889/checks?check_run_id=1903151385

I switched this over to a once-a-day schedule.

File: hack/netpol-cyclonus/install-cyclonus.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+apiVersion: batch/v1
```
Comment (aanm) [inline hack/netpol-cyclonus/install-cyclonus.yml]: Perhaps we should move this into `test/fuzzing`?

File: hack/netpol-cyclonus/test-cyclonus.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
```
Comment (aanm) [inline hack/netpol-cyclonus/test-cyclonus.sh]: Perhaps we should move this into `test/fuzzing`?

**Review by aanm — CHANGES_REQUESTED** [review]
@twpayne what do you think, should we move this to `test/fuzzing`?

File: hack/netpol-cyclonus/install-cyclonus.yml (hunk 1 above)
Comment (twpayne) [reply hack/netpol-cyclonus/install-cyclonus.yml]: This definitely deserves its own directory. Cyclonus does proper end-to-end testing of network policies, it's not a fuzzer, so I would put it in `test/cyclonus` or `test/networkpolicy` or similar.

Comment (mattfenwick) [reply hack/netpol-cyclonus/install-cyclonus.yml]: 👍 👍 

Hopefully we can get cyclonus to doing something more similar to fuzzing in the not-too-distant future ? 😄   I think part of the calculus on where it goes next it what provides the most value to cilium (regression tests?  conformance tests?  tests against randomly generated network policies?  something else ... 🤔 )  

Whatever does end up being the best route, I think the changes in this github workflow will be tiny: bump cyclonus to a new version, change the args passed in -- shouldn't be disruptive or complicated!

Anyway just 🤔 out loud here, happy to collab more on this!  (Not looking to just do a drive-by and dump some quickly out-of-date tests on you 😄 😄 )

**Comment by twpayne** [comment]
test-me-please

File: .github/workflows/cyclonus-netpol-test.yaml [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+      - name: Precheck generated connectivity manifest files
+        run: |
+          make -C examples/kubernetes/connectivity-check fmt
+          make -C examples/kubernetes/connectivity-check all
+          git diff --exit-code || (echo "please run 'make -C examples/kubernetes/connectivity-check all' and submit your changes"; exit 1)
+
+      - name: Build docker images
+        run: |
+          make docker-cilium-image
+          make docker-operator-generic-image
```
Comment (aanm) [inline .github/workflows/cyclonus-netpol-test.yaml]: The images are no longer built this way, please check how this file is doing:
https://github.com/cilium/cilium/blob/master/.github/workflows/kind-1.19.yaml#L21-L28

https://github.com/cilium/cilium/blob/master/.github/workflows/kind-1.19.yaml#L42-L47

**Review by aanm — CHANGES_REQUESTED** [review]

**Review by twpayne — APPROVED** [review]
LGTM once @aanm's comments are addressed.

File: .github/workflows/cyclonus-netpol-test.yaml (hunk 3 above)
Comment (mattfenwick) [reply .github/workflows/cyclonus-netpol-test.yaml]: 👍 okay I just removed that step and copy/pasted the other two steps in its place, 🤞 that will work !

File: hack/netpol-cyclonus/install-cyclonus.yml (hunk 1 above)
Comment (twpayne) [reply hack/netpol-cyclonus/install-cyclonus.yml]: @mattfenwick would you mind moving these files to a `test/netpol-cyclonus` directory so they're alongside the other tests? Thanks!

File: .github/workflows/cyclonus-netpol-test.yaml [hunk 4]
```diff
@@ -0,0 +1,114 @@
+name: Cyclonus network policy test
+
+on:
+  # TODO remove push after it runs successfully
```
Comment (mattfenwick) [inline .github/workflows/cyclonus-netpol-test.yaml]: TODO remove push after it runs successfully

**Review by nbusseneau — APPROVED** [review]
LGTM as well.

**Comment by mattfenwick** [comment]
Trying to get this job to run successfully before removing the `pull_request` and `push` triggers!

**Comment by mattfenwick** [comment]
Looks like the image isn't getting pushed where the job expects it, have to dig in to figure out why

File: .github/workflows/cyclonus-netpol-test.yaml [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        uses: helm/kind-action@v1.1.0
+        with:
+          version: ${{ env.KIND_VERSION }}
+          config: ${{ env.KIND_CONFIG }}
+
+      - name: Load local images into kind cluster
+        run: |
+          docker images
+          kind load docker-image --name chart-testing quay.io/cilium/cilium:latest
+          kind load docker-image --name chart-testing quay.io/cilium/operator-generic:latest
```
Comment (aanm) [inline .github/workflows/cyclonus-netpol-test.yaml]: This can be removed.

File: .github/workflows/cyclonus-netpol-test.yaml [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+             --set hostServices.enabled=false \
+             --set externalIPs.enabled=true \
+             --set nodePort.enabled=true \
+             --set hostPort.enabled=true \
+             --set bpf.masquerade=false \
+             --set ipam.mode=kubernetes \
+             --set image.tag=latest \
+             --set image.pullPolicy=Never \
+             --set operator.image.tag=latest \
+             --set operator.image.pullPolicy=Never \
```
Comment (aanm) [inline .github/workflows/cyclonus-netpol-test.yaml]: This needs to be set as such:
```
             --set image.repository=quay.io/${{ github.repository_owner }}/cilium-ci \
             --set image.tag=${{ steps.vars.outputs.tag }} \
             --set image.pullPolicy=IfNotPresent \
             --set operator.image.repository=quay.io/${{ github.repository_owner }}/operator \
             --set operator.image.suffix=-ci \
             --set operator.image.tag=${{ steps.vars.outputs.tag }} \
             --set operator.image.pullPolicy=IfNotPresent \
```

**Review by aanm — CHANGES_REQUESTED** [review]
@mattfenwick I've left some comments that will help you unblock

File: .github/workflows/cyclonus-netpol-test.yaml (hunk 5 above)
Comment (mattfenwick) [reply .github/workflows/cyclonus-netpol-test.yaml]: 👍

File: .github/workflows/cyclonus-netpol-test.yaml (hunk 6 above)
Comment (mattfenwick) [reply .github/workflows/cyclonus-netpol-test.yaml]: 👍

**Comment by mattfenwick** [comment]
Thanks @aanm !!  made those changes and giving it another whirl now 😄

**Comment by mattfenwick** [comment]
🎉 looks like it worked!!  🎉

File: .github/workflows/cyclonus-netpol-test.yaml (hunk 4 above)
Comment (aanm) [reply .github/workflows/cyclonus-netpol-test.yaml]: I assume this can be removed now?

**Review by aanm — COMMENTED** [review]
Thanks @mattfenwick!

Also, can you squash all commits together? Thanks!

File: .github/workflows/cyclonus-netpol-test.yaml (hunk 4 above)
Comment (mattfenwick) [reply .github/workflows/cyclonus-netpol-test.yaml]: 👍

**Comment by mattfenwick** [comment]
Squashed 🎉 😄

**Comment by twpayne** [comment]
Thanks @mattfenwick - this is fantastic!

**Review by aanm — APPROVED** [review]

**Pull request merged**